### DISCUSSION TITLE
The firmware breaks after the line "Erasing flash (this may take a while)..." esp8266

### DISCUSSION THREAD

**Discussion by kalabaha2022** [micropython/micropython#9774] [General]
Hello everyone. I have a problem with lolin nodemcu v3 board. When you try to erase the flash or write the firmware, the process ends after the line "Erasing flash (this may take a while) ...". I figured there might be a problem with a defective flush. I soldered it and tried to read it, the flash turned out to be working. I tried to remove the dump from the working board and sew it into a flash through the programmer. Nothing helped
> esptool.py v4.3
Found 1 serial ports
Serial port /dev/ttyUSB0
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: d8:bf:c0:cd:89:bc
Uploading stub...
Running stub...
Stub running...
Erasing flash (this may take a while)...
Traceback (most recent call last):
  File "/home/user/.local/bin/esptool.py", line 34, in <module>
    esptool._main()
  File "/home/user/.local/lib/python3.9/site-packages/esptool/__init__.py", line 1022, in _main
    main()
  File "/home/user/.local/lib/python3.9/site-packages/esptool/__init__.py", line 824, in main
    operation_func(esp, args)
  File "/home/user/.local/lib/python3.9/site-packages/esptool/cmds.py", line 984, in erase_flash
    esp.erase_flash()
  File "/home/user/.local/lib/python3.9/site-packages/esptool/loader.py", line 109, in inner
    return func(*args, **kwargs)
  File "/home/user/.local/lib/python3.9/site-packages/esptool/loader.py", line 1049, in erase_flash
    self.check_command(
  File "/home/user/.local/lib/python3.9/site-packages/esptool/loader.py", line 407, in check_command
    val, data = self.command(op, data, chk, timeout=timeout)
  File "/home/user/.local/lib/python3.9/site-packages/esptool/loader.py", line 376, in command
    p = self.read()
  File "/home/user/.local/lib/python3.9/site-packages/esptool/loader.py", line 308, in read
    return next(self._slip_reader)
StopIteration

Also, when the board is connected in the terminal on baud 74600, the line "ets Jan 8 2013,rst cause:2, boot mode:(3,6) appears
"
I tried to flash this board on different devices, the same problems. I also successfully flashed other esp and arduino on my device
I would like to understand the reason, since this is not the first board with such symptoms for me.
P.S.
Sorry for bad English

P.S.
I also connected pin reset and gng to prevent esp from starting. I connected RX->TX, then sent a message through the serial port to check the operation of Ch340. It works well

**Comment by robert-hh** [comment]
Are you sure that the board is i bootloader mode when you try to load the firmware. The boot message does not show that. For bootloader mode, GPIO0 must be connected to GND and then Reset must be pushed, or power must by cycled.
At reset, you should at 74800 see a message like `ets Jan  8 2013,rst cause:2, boot mode:(1,6)`

**Reply by kalabaha2022** [reply]
Yes, I did this and got a similar record, but this does not affect the process of flashing the board

**Reply by robert-hh** [reply]
You could try to set chip type and flash size manually with the esptool options --chip esp8266 --flash_size 4MB:

`esptool.py --port /dev/ttyUSB0 --baud 460800 --chip esp8266 write_flash --flash_size 4MB 0 <firmware file name>`

**Reply by kalabaha2022** [reply]
command execution ends like this: `A fatal error occurred: No serial data received.`

**Reply by robert-hh** [reply]
That's strange. So you have the device at /dev/ttyUSB0? Any other output in between besides that message?

**Reply by kalabaha2022** [reply]
The device is detected. The full output of the command line when executing the command is as follows:

> esptool.py --port /dev/ttyUSB0 --baud 460800 --chip esp8266 write_flash --flash_size 4MB 0 123.bin
esptool.py v4.3
Serial port /dev/ttyUSB0
Connecting....
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: d8:bf:c0:cd:89:bc
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
A fatal error occurred: No serial data received.

I hope I understood correctly what you mean)

**Reply by robert-hh** [reply]
Yes. So it seems that the stub is not responding. Is 123.bin a proper firmware image for loading?

**Reply by kalabaha2022** [reply]
I tested this image on a similar board, also tried other images. Perhaps I came across some special revision of the board where there are some peculiarities or just a marriage

**Reply by robert-hh** [reply]
You could try something simple to check, whether the stub works:

`esptool.py --port /dev/ttyUSB0 --baud 460800 --chip esp8266 chip_id`

That should return the chip ID. Example:
```
esptool.py v3.3-dev
Serial port /dev/ttyUSB0
Connecting....
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: 5c:cf:7f:85:ea:1a
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Chip ID: 0x0085ea1a
Hard resetting via RTS pin...
```
Comparing I see, that in your log not even the baud rate change is confirmed. So there may be a problem aside the ESP8266 board.

**Reply by robert-hh** [reply]
You could try as well the lower baud rate:
`esptool.py --port /dev/ttyUSB0 --baud 115200 --chip esp8266 chip_id`

**Reply by kalabaha2022** [reply]
460800:
> esptool.py --port /dev/ttyUSB0 --baud 460800 --chip esp8266 chip_id
esptool.py v4.3
Serial port /dev/ttyUSB0
Connecting....
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: d8:bf:c0:cd:89:bc
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
A fatal error occurred: No serial data received.

115200:
>  esptool.py --port /dev/ttyUSB0 --baud 115200 --chip esp8266 chip_id
esptool.py v4.3
Serial port /dev/ttyUSB0
Connecting....
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: d8:bf:c0:cd:89:bc
Uploading stub...
Running stub...
Stub running...
Traceback (most recent call last):
  File "/home/user/.local/bin/esptool.py", line 34, in <module>
    esptool._main()
  File "/home/user/.local/lib/python3.9/site-packages/esptool/__init__.py", line 1022, in _main
    main()
  File "/home/user/.local/lib/python3.9/site-packages/esptool/__init__.py", line 824, in main
    operation_func(esp, args)
  File "/home/user/.local/lib/python3.9/site-packages/esptool/cmds.py", line 966, in chip_id
    chipid = esp.chip_id()
  File "/home/user/.local/lib/python3.9/site-packages/esptool/targets/esp8266.py", line 130, in chip_id
    id0 = self.read_reg(self.ESP_OTP_MAC0)
  File "/home/user/.local/lib/python3.9/site-packages/esptool/loader.py", line 690, in read_reg
    val, data = self.command(
  File "/home/user/.local/lib/python3.9/site-packages/esptool/loader.py", line 376, in command
    p = self.read()
  File "/home/user/.local/lib/python3.9/site-packages/esptool/loader.py", line 308, in read
    return next(self._slip_reader)
StopIteration

**Reply by robert-hh** [reply]
Maybe a version problem of esptool.py You could try the version I used. You have to make it executable.
[esptool.zip](https://github.com/micropython/micropython/files/9893932/esptool.zip)

**Reply by robert-hh** [reply]
Version 4.3 works here as well.

**Reply by robert-hh** [reply]
You could try the no-stub option to see, if it is a problem with the stub not running.
`esptool.py --port /dev/ttyUSB0 --baud 115200 --no-stub --chip esp8266 chip_id`

**Reply by kalabaha2022** [reply]
With esptool 3.3 the result is the same. When using the --no-stub:

esptool.py --port /dev/ttyUSB0 --baud 115200 --no-stub --chip esp8266 chip_id
esptool.py v4.3
Serial port /dev/ttyUSB0
Connecting....
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: d8:bf:c0cd89:bc
Enabling default SPI flash mode...
A fatal error occurred: Serial data stream stopped: Possible serial noise or corruption.

**Reply by robert-hh** [reply]
So it's a serial communication problem. Which serial module for python do you have installed with pip. Is it pyserial?

**Reply by robert-hh** [reply]
Maybe teh baud rate cannot be reconfigured. Since you have some communication at 74880 baud, you could try:
`esptool.py --port /dev/ttyUSB0 --baud 74880 --chip esp8266 --no-stub chip_id` or
`esptool.py --port /dev/ttyUSB0 --baud 74880 --chip esp8266 chip_id`

**Reply by kalabaha2022** [reply]
On 74880 the result is the same. I'm not sure about my serial module. How can I see which serial module is being used?

**Reply by robert-hh** [reply]
`pip list`
should show a list of installed modules.
`pip install pyserial` will install the pyserial module.

**Reply by kalabaha2022** [reply]
pyserial has already been installed

**Reply by robert-hh** [reply]
Can you take a picture of your board's top side, as detailed as possible?

**Reply by kalabaha2022** [reply]
I forgot to clarify that other firmware software (flash download tools, espflasher a. o.) did not help solve the problem

**Reply by robert-hh** [reply]
And just as a another blind guess, you could try a different USB cable.

**Reply by kalabaha2022** [reply]
I have already changed several different cables) I am taking a photo now

**Reply by kalabaha2022** [reply]
![photo_2022-10-29_19-19-22](https://user-images.githubusercontent.com/116905801/198842245-adc2275b-3328-49fe-918a-b6dd72bb18bc.jpg)
![photo_2022-10-29_19-19-28](https://user-images.githubusercontent.com/116905801/198842250-ca84c2ca-5204-4832-8781-4ca4ba2ea28e.jpg)

**Reply by robert-hh** [reply]
The board looks OK. The mdoe switch transistors are populated. Next step for me would be logging, what actually happens on the wire. I do not know if you can do that. At least you can make a loopback test by:
- Connecting RST to GND, and
- Connecting RX with TX at the board.

Then all characters should be echoed. Then you know, that at least the USB/UART bridge works.

**Reply by kalabaha2022** [reply]
Already did today and supplemented my post at the top. Characters are returned without problems and at any speed

**Reply by robert-hh** [reply]
ok. did you check if there is a connection between rx, tx and the esp8266 chip pins are good?

**Reply by kalabaha2022** [reply]
And from here strangeness. There is no direct contact between rx, tx between board and chip. I found a diagram of my board and checked the contacts on it. there is more than 10kOm on the tx contact between the board and the chip, there is no contact on the rx

**Reply by kalabaha2022** [reply]
After researching, I found the right pins. Between the chip and the esp at rx, tx 200 ohm resistor. And pinout does not comply with ch340G. TX - 10, RX - 14, GNG - 7, 3V - 11

**Comment by robert-hh** [comment]
I did a capture of RX TX of the chip-id command. The data itself does not tell me a lot, but at least one can see the timing. The first pic shows the full transaction, starting with the reset-response. The second & third pic show a zooms starting with the chip-id command communication. You see a clear alternating communication, which is always in time.

![full_sample](https://user-images.githubusercontent.com/12476868/198848777-4193d3e8-15ce-4713-8e51-5e20dc1b4908.jpg)

![zoom_sample](https://user-images.githubusercontent.com/12476868/198848782-8083fa5f-6c40-4cea-bc7c-5f5cfed0f697.jpg)

![Zoom_bytes](https://user-images.githubusercontent.com/12476868/198848936-06808e9c-594b-411f-884b-05fde857d750.jpg)

**Reply by kalabaha2022** [reply]
Alas, I do not have the equipment to check the signals. I'll try to remove the chip or remove the esp12 from the board, and flash it through another usb-uart converter. If this does not help, then the problem is exactly inside the esp8266 microcontroller.

**Reply by robert-hh** [reply]
I agree that the problem is most likely in the ESP12 module. The ESP8266 starts. Otherwise it would not send that boot message. It still puzzles me, that you do not get the `boot mode:(1,6)` line, when GPIO0 is at GND and reset is pushed. Because that is the boot loader state. `boot mode:(3,6)` is the normal boot state. 

It may be easier to just get another ESP8266 board. I usually recommend to get a simple logic analyzer  tool. There are devices are  available at about 10€ with a sample rate of 24MHz, which work with the Saleae logic software or the PulseView software. For serial communication analysis they a worth every cent. Available e.g. at aliexpress or Amazon or others. Like: https://www.aliexpress.com/item/1005003812410259.html?spm=a2g0o.productlist.0.0.20ee7ba1RG5MCU&algo_pvid=f3a6dcf9-88be-4e5e-b6f5-37437a90ab0d&algo_exp_id=f3a6dcf9-88be-4e5e-b6f5-37437a90ab0d-0&pdp_ext_f=%7B%22sku_id%22%3A%2212000027249902447%22%7D&pdp_npi=2%40dis%21EUR%217.26%216.25%21%21%21%21%21%402100bddb16671143546966380e61f6%2112000027249902447%21sea&curPageLogUid=IjO3DzklpdfV

**Reply by kalabaha2022** [reply]
Thank you for the recommendation for the recommendation, I will take into account this device and purchase it if possible. Also, thank you very much for your attention to my question, and the help you rendered to me.
